Fix `decode_ignored_any`

DrewMcArthur opened the pull request:
this beefs up the implementation for ignoring values, mostly lifted from serde's implementation.  added a bunch of tests, which pass and these changes also fix the broken tests that led me to these changes in another repository of mine that's making use of this library.

- would appreciate a closer look / double checking my implementations of some of the more base level helper functions here (e.g. `peek_or_null`, `eat_char`, etc.)
- definitely take a look at the tests to see if there's anything i missed there that you'd like to cover
- i think there's still some cleanup i can do; i'll take another pass at this later today.

DrewMcArthur commented:
closes #2

DrewMcArthur commented on src/de.rs:
```diff
@@ -398,6 +400,50 @@ impl<S: Read> Decoder<S> {
         Ok(i)
     }
 
+    pub(crate) async fn peek(&mut self) -> Result<Option<u8>, Error> {
```
```suggestion
    async fn peek(&mut self) -> Result<Option<u8>, Error> {
```
might not need this level visibility

haydnv replied:
Yes there's no decoding logic outside this module so I think this should be private

DrewMcArthur commented on src/de.rs:
```diff
@@ -452,23 +498,101 @@ impl<S: Read> Decoder<S> {
-        if self.buffer.is_empty() {
-            Ok(())
-        } else {
-            if self.buffer.starts_with(QUOTE) {
-                self.parse_string().await?;
-            } else if self.numeric.contains(&self.buffer[0]) {
-                self.parse_number::<f64>().await?;
-            } else if self.buffer[0] == b'n' {
-                self.parse_unit().await?;
-            } else {
-                self.parse_bool().await?;
```
i didn't check if any of this results in dead code that can be deleted or not

DrewMcArthur replied:
assuming clippy would catch that, so resolving

DrewMcArthur commented on src/de.rs:
```diff
@@ -452,23 +498,101 @@ impl<S: Read> Decoder<S> {
+
+    async fn ignore_string(&mut self) -> Result<(), Error> {
+        // eat the first char, which is a quote
+        self.next_or_eof().await?;
```
```suggestion
        self.eat_char().await?;
```

DrewMcArthur replied:
these functions basically do the same, it's just slightly clearer what's happening

DrewMcArthur commented on src/lib.rs:
```diff
@@ -428,6 +428,60 @@ mod tests {
+            }
+        }
+
+        let encoded = "{
+            \"string\": \"Hello, world!\",
+            \"number\": 42,
+            \"boolean\": true,
+            \"array\": [1, 2, 3],
+            \"object\": {\"key\": \"value\"},
+            \"null_value\": null,
+            \"nested_object\": {
+              \"nested_string\": \"Nested string\",
+              \"nested_number\": 3.14,
+              \"nested_boolean\": false,
+              \"nested_array\": [\"apple\", \"banana\", \"orange\"],
+              \"nested_null\": null
+            },
+            \"unicode_characters\": \"💡🌟🔑\",
+            \"empty_array\": [],
+            \"empty_object\": {},
+            \"multiline_string\": \"This is a\\nmultiline\\nstring.\",
+            \"escaped_characters\": \"Escaped characters: \\\" \\\\ \\/ \\b \\f \\n \\r \\t \\u1234\"
+          }";
+
+        let source = stream::iter(encoded.as_bytes().iter().copied())
+            .chunks(2)
+            .map(Bytes::from);
+
+        let _: IgnoredValue = decode((), source).await.unwrap();
```
should we also put this into a `test_file.json` and read from that, to check that works as expected, without having to deal with escaping `"` and whatnot?

haydnv replied:
I think this would work fine as a raw string: 
```rust
let encoded = r#"{"string": "Hello, World!", ...}"#;
```

here are some examples: https://rahul-thakoor.github.io/rust-raw-string-literals/

DrewMcArthur commented on src/de.rs:
```diff
@@ -895,3 +1150,167 @@ pub async fn read_from<S: AsyncReadExt + Send + Unpin, T: FromStream>(
+
+    #[tokio::test]
+    async fn test_ignore_number() {
+        let inputs = [
+            ("0", 0),
+            ("123", 0),
+            ("-123", 0),
+            ("123.45", 0),
+            ("-123.45", 0),
+            ("0.0", 0),
+            ("123, 45", 4),
+        ];
+
+        for input in inputs {
```
should probably move these into `#[test_case(...)]`s like the following two tests

DrewMcArthur commented on src/de.rs:
```diff
@@ -895,3 +1150,167 @@ pub async fn read_from<S: AsyncReadExt + Send + Unpin, T: FromStream>(
+    /// ignore_exactly takes a vector of bytes, and consumes exactly those characters.
+    #[tokio::test]
+    async fn test_ignore_exactly() {
+        let inputs = [("foo", "foo", true, 0), ("foobar", "foo", true, 3)];
```
should add more cases to this test, ensure it fails when we want it to

DrewMcArthur replied:
added one more case; let me know if we should have more

DrewMcArthur commented on src/de.rs:
```diff
@@ -556,6 +680,137 @@ impl<S: Read> Decoder<S> {
+                    self.eat_char().await?;
+                    return Ok(());
+                }
+                b',' => {
+                    self.ignore_exactly(",").await?;
+                }
+                _ => {
+                    self.ignore_value().await?;
+                }
```
this should probably be just one arm of the `match`.  i think there would be a bug for
`["test""test"]` would be allowed, when that isn't valid json.

```suggestion
                _ => {
                    self.ignore_value().await?;
                    self.expect_whitespace().await?;
                    self.ignore_exactly(",").await?;
                }
```

DrewMcArthur replied:
i think i've fixed this, but haven't put in any tests expecting errors yet

DrewMcArthur commented on src/de.rs:
```diff
@@ -556,6 +680,137 @@ impl<S: Read> Decoder<S> {
+                    self.eat_char().await?;
+                    return Ok(());
+                }
+                b',' => self.eat_char().await?,
+                _ => {
+                    self.ignore_string().await?; // key
+                    self.expect_whitespace().await?;
+                    self.ignore_exactly(":").await?;
+                    self.ignore_value().await?;
+                }
```
same as above.  should add a test case for this
```suggestion
                _ => {
                    self.ignore_string().await?; // key
                    self.expect_whitespace().await?;
                    self.ignore_exactly(":").await?;
                    self.ignore_value().await?;
                    self.expect_whitespace().await?;
                    self.ignore_exactly(",").await?;
                }
```

haydnv commented on src/de.rs:
```diff
@@ -398,6 +400,50 @@ impl<S: Read> Decoder<S> {
+    }
+
+    async fn peek_or_null(&mut self) -> Result<u8, Error> {
+        Ok(self.peek().await?.unwrap_or(b'\x00'))
```
I think it would be better to check if `peek` returns `Some(...)` since AFAIK the null byte has no significance in the JSON spec

DrewMcArthur replied:
these methods i copied over from serde - i think you're right though.  i've removed both of these methods, and refactored the usages to just use `peek()` or `next_char()`

haydnv commented on src/de.rs:
```diff
@@ -398,6 +400,50 @@ impl<S: Read> Decoder<S> {
+    }
+
+    async fn next_char_or_null(&mut self) -> Result<u8, Error> {
+        Ok(self.next_char().await?.unwrap_or(b'\x00'))
```
Same here, wouldn't it be better to check if `next_char` is `None` than to return a null?

haydnv commented on src/de.rs:
```diff
@@ -452,23 +498,101 @@ impl<S: Read> Decoder<S> {
+    async fn ignore_string(&mut self) -> Result<(), Error> {
+        // eat the first char, which is a quote
+        self.next_or_eof().await?;
+        loop {
```
nit: this would be clearer with more newlines

haydnv reviewed (commented):
Thanks Drew! Do you think you could bring some of the parsing improvements into the `parse_*` methods? For example `parse_number` could call `parse_decimal`, `parse_exponent`, etc. This could be in a separate PR if you prefer.

DrewMcArthur commented:
@haydnv thanks for the review - 
1. i opened #5 for refactoring the parse methods.  i'd be open to it, but first would like to implement some benchmark tests to make sure any changes only speed things up!
2. I think I addressed all the comments.  would appreciate another glance to make sure everything is covered.
3. only things missing are some "expected failures" tests.  i can add those here if you'd like, or we can open an issue to make sure they eventually get added.

haydnv commented on src/de.rs:
```diff
@@ -4,7 +4,9 @@ use std::collections::HashSet;
 
+use async_recursion::async_recursion;
 use async_trait::async_trait;
+
```
nit: remove extra newline

DrewMcArthur replied:
```suggestion
```

haydnv commented:
It would be great if you can add the expected failure tests in this PR

DrewMcArthur commented:
okay! @haydnv sorry for the delay.  i've added a bunch more test cases, and ran `cargo llvm-cov --features all --output-path lcov.info --lcov` to verify that all the code paths in these `ignore_...` functions are covered.

haydnv commented:
Thanks Drew, great work!
